### DISCUSSION TITLE
Changelog for NPM packages hosted in AWS CodeArtifact

### DISCUSSION THREAD

**Discussion by otbe** [renovatebot/renovate#26721] [Request Help]
### How are you running Renovate?

Self-hosted Renovate

### If you're self-hosting Renovate, tell us what version of Renovate you run.

37.129.1

### If you're self-hosting Renovate, select which platform you are using.

gitlab.com

### Was this something which used to work for you, and then stopped?

I have not seen this working

### Describe the problem

Im trying to get changelog/releases working for a NPM package in a private Gitlab(.com) repository and published to AWS CodeArtifact.

My package.json looks like similar to this:
```
{
  "name": "@my-scope/my-package",
  "version": "0.2.1",
  "main": "index.js",
  "scripts": {
    "...": "..."
  },
  "description": "FOO description",
  "keywords": [],
  "author": {
    "name": "ME",
    "email": "Me@domain.tld"
  },
  "repository": {
    "type": "git",
    "url": "https://gitlab.com/my-group/my-repo.git"
  },
  "license": "ISC",
  "stability": "stable",
  "types": "index.d.ts",
  "devDependencies": {

  },
  "peerDependencies": {

  }
}
```

If I look at the package from a npm perspective via `npm view @my-scope/my-package --json` it looks like this:
```
{
  "_id": "@my-scope/my-package@0.2.1",
  "name": "@my-scope/my-package",
  "description": "FOO description",
  "versions": [
    "0.2.0",
    "0.2.1-dev.5953239959",
    "0.2.1"
  ],
  "access": null,
  "dist-tags": {
    "latest": "0.2.1",
    "next": "0.2.1-dev.5953239959"
  },
  "time": {
    "modified": "2024-01-17T21:30:11.747Z",
    "0.2.0": "2024-01-17T20:08:27.668Z",
    "0.2.1": "2024-01-17T21:27:57.927Z",
    "0.2.1-dev.5953239959": "2024-01-17T21:29:05.009Z"
  },
  "_contentLength": 0,
  "version": "0.2.1",
  "main": "index.js",
  "scripts": {

  },
  "keywords": [],
  "author": "ME <Me@domain.tld>",
  "repository": {
    "type": "git",
    "url": "git+https://gitlab.com/my-group/my-repo.git"
  },
  "license": "ISC",
  "stability": "stable",
  "types": "index.d.ts",
  "devDependencies": {

  },
  "peerDependencies": {

  },
  "readme": "# README",
  "readmeFilename": "README.md",
  "bugs": {
    "url": "https://gitlab.com/my-group/my-repo/issues"
  },
  "homepage": "https://gitlab.com/my-group/my-repo#readme",
  "_integrity": "",
  "_resolved": "",
  "_from": "",
  "_nodeVersion": "18.18.2",
  "_npmVersion": "8.19.4",
  "dist": {
    "integrity": "sha512-",
    "shasum": "....",
    "tarball": "https://mydomain-123456788.d.codeartifact.eu-central-1.amazonaws.com/....."
  }
}
```

A maven project also hosted in Gitlab and published to CodeArtifact works perfectly fine via 
```
<scm>
  <connection>scm:git:git@gitlab.com:my-group/my-repo2.git</connection>
  <developerConnection>scm:git:git@gitlab.com:my-group/my-repo2.git</developerConnection>
  <url>https://gitlab.com/my-group/my-repo2.git</url>
  <tag>HEAD</tag>
</scm>
```

### Relevant debug logs

<details><summary>Logs</summary>

Renovate reports updates to the package like this
```

{
  "currentValue": "0.1.2",
  "currentVersion": "0.1.2",
  "datasource": "npm",
  "depName": "@my-scope/my-package",
  "depType": "devDependencies",
  "fixedVersion": "0.1.2",
  "isSingleVersion": true,
  "lockedVersion": "0.1.2",
  "packageName": "@my-scope/my-package",
  "prettyDepType": "devDependency",
  "registryUrl": "https://mydomain-123456788.d.codeartifact.eu-central-1.amazonaws.com/....",
  "versioning": "npm",
  "warnings": [],
  "updates": [
    {
      "bucket": "non-major",
      "newVersion": "0.2.1",
      "newValue": "0.2.1",
      "releaseTimestamp": "2024-01-17T21:27:57.927Z",
      "newMajor": 0,
      "newMinor": 2,
      "updateType": "minor",
      "branchName": "renovate/my-scope-my-package-0.x"
    }
  ]
}

```
What I notice here is that theres no "sourceUrl" included. Renovate bot is not able to detect the source somehow.
</details>

### Have you created a minimal reproduction repository?

I have explained in the description why a minimal reproduction is impossible

**Comment by rarkins** [comment]
First step is working out why the sourceUrl is not detected by Renovate. Unfortunately this may only be possible through debugging Renovate, which only you can do if it's a private package

**Reply by otbe** [reply]
Could you point my to a test where the sourceUrl is extracted from the registry response or something like this? Then I can start looking into this :)

**Reply by rarkins** [reply]
https://github.com/search?q=repo%3Arenovatebot%2Frenovate%20path%3A%2F%5Elib%5C%2Fmodules%5C%2Fdatasource%5C%2Fnpm%5C%2F%2F%20sourceurl&type=code

**Reply by otbe** [reply]
Ok I was able to track down the issue.

I enabled trace logging for one affected project and found the actual sourceUrl renovate discovers. It looks like:
`git+ssh://foo:bar@gitlab.com/my/group/repo.git`

In [metadata.ts](https://github.com/renovatebot/renovate/blob/19eb68b6a6e422e5d508b203c8c4adb1e54760fe/lib/modules/datasource/metadata.ts#L167-L178) after `massageUrl` it looks like `git+ssh://foo:bar@gitlab.com/my/group/repo` and this is then passed to [github-url-from-git](https://www.npmjs.com/package/github-url-from-git) with an additional host for gitlab. 

However this returns undefined and the [validateUrl](https://github.com/renovatebot/renovate/blob/19eb68b6a6e422e5d508b203c8c4adb1e54760fe/lib/modules/datasource/metadata.ts#L191C33-L191C33) call later deletes the source url. 

[github-url-from-git](https://www.npmjs.com/package/github-url-from-git) returns undefined because it's not able to parse nested groups in the url. A flat layout like "group/repo" works also for gitlab. See https://runkit.com/embed/054bust0ses8

(foo:bar is redacted in the actual logs, no worries :D)

**Reply by rarkins** [reply]
Great troubleshooting! Unfortunately that means that github-url-from-git needs enhancing, or we need to replace it entirely with our own logic.

**Reply by rarkins** [reply]
They may not want to enhance it for nested groups considering it's a _github_ lib

**Reply by otbe** [reply]
Yes especially after this issue was already raised several years ago: https://github.com/tj/node-github-url-from-git/issues/23

Not quite sure how to proceed. While looking around I found https://www.npmjs.com/package/parse-git-url  which looks promising but with a different API. Especially the part with additional hosts might get complicated there.

**Reply by rarkins** [reply]
We could try falling back to multiple packages if none are perfect

**Reply by otbe** [reply]
An easy fix could be to use [parse-git-url](https://www.npmjs.com/package/parse-git-url) after [github-url-from-git](https://www.npmjs.com/package/github-url-from-git) similar to ` || dep.sourceUrl`. Wdyt? 

This addition would work without breaking anything serious and solve the issue with nested gitlab.com groups at least 🤔

**Reply by otbe** [reply]
Created https://github.com/renovatebot/renovate/pull/26766
